// astro-compressor/dist/compress.js
import { createReadStream, createWriteStream } from "node:fs";
import { readdir } from "node:fs/promises";
import { extname, resolve } from "node:path";
import { hrtime } from "node:process";
import { promises as stream } from "node:stream";
import * as zlib from "node:zlib";
export async function* walkDir(dir, extensions) {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
        const name = resolve(dir, entry.name);
        if (entry.isDirectory()) {
            yield* walkDir(name, extensions);
        }
        else if (filterFile(entry.name, extensions)) {
            yield name;
        }
    }
}
const filterFile = (file, extensions) => {
    return extensions.some((ext) => extname(file) === ext);
};
const compress = async (name, compressedFileNames, compressor, logger, { files, batchSize, enabled, options }) => {
    if (!enabled) {
        logger.warn(`${name} compression disabled, skipping...`);
        return;
    }
    const start = hrtime.bigint();
    for (let i = 0; i < files.length; i += batchSize) {
        const batch = files.slice(i, i + batchSize);
        await Promise.all(batch.map(async (path) => {
            const source = createReadStream(path);
            const destination = createWriteStream(`${path}.${compressedFileNames}`);
            const comp = compressor(options);
            await stream.pipeline(source, comp, destination);
        }));
    }
    const end = hrtime.bigint();
    logger.info(`${name.padEnd(8, " ")} compressed ${files.length} files in ${(end - start) / BigInt(1000000)}ms`);
};
export const gzip = async (files, logger, enabled, batchSize = 10) => {
    await compress("gzip", "gz", zlib.createGzip, logger, {
        files,
        enabled: enabled === true || typeof enabled === "object",
        options: typeof enabled === "object" ? enabled : { level: zlib.constants.Z_BEST_COMPRESSION },
        batchSize,
    });
};
export const brotli = async (files, logger, enabled, batchSize = 10) => {
    await compress("brotli", "br", zlib.createBrotliCompress, logger, {
        files,
        enabled: enabled === true || typeof enabled === "object",
        options: typeof enabled === "object" ? enabled : undefined,
        batchSize,
    });
};
export const zstd = async (files, logger, enabled, batchSize = 10) => {
    if (typeof zlib.createZstdCompress !== "function") {
        logger.warn("zstd compression is not supported in this Node.js version.");
        return;
    }
    await compress("zstd", "zst", zlib.createZstdCompress, logger, {
        files,
        enabled: enabled === true || typeof enabled === "object",
        options: typeof enabled === "object" ? enabled : undefined,
        batchSize,
    });
};
