// astro-compressor/dist/index.js
import { fileURLToPath } from "node:url";
import { brotli, gzip, walkDir, zstd } from "./compress.js";
const defaultFileExtensions = [".css", ".js", ".html", ".xml", ".cjs", ".mjs", ".svg", ".txt"];
const defaultOptions = {
    gzip: true,
    brotli: true,
    zstd: true,
    fileExtensions: defaultFileExtensions,
    batchSize: 10,
};
export default function (opts = defaultOptions) {
    const options = { ...defaultOptions, ...opts };
    return {
        name: "astro-compressor",
        hooks: {
            "astro:build:done": async ({ dir, logger }) => {
                const path = fileURLToPath(dir);
                const files = [];
                for await (const file of walkDir(path, options.fileExtensions)) {
                    files.push(file);
                }
                await Promise.allSettled([
                    gzip(files, logger, options.gzip, options.batchSize),
                    brotli(files, logger, options.brotli, options.batchSize),
                    zstd(files, logger, options.zstd, options.batchSize),
                ]);
                logger.info("Compression finished\n");
            },
        },
    };
}
